### PR TITLE
Convert gifs to videos for faster loading

### PR DESCRIPTION
**Pull request by matthewdias**

### CODE REVIEW

File: lib/html/pipeline/gif_filter.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+      def to_link(url)
+        case url
+        when %r{https?://.*\.imgur.com/([^\?]+).gifv?}
+          "https://imgur.com/#{$1}"
+        # this is useless until giphy supports embedding video
+        when %r{https?://.*\.giphy.com/media/([^\?]+)/giphy.gif}
+          "https://giphy.com/gifs/#{$1}"
+        when %r{https?://.*\.gifs.com/([^\?]+).gif}
+          "https://gifs.com/gif/#{$1}"
```
[the remark this answers is not in the document]
Comment (matthewdias) [reply lib/html/pipeline/gif_filter.rb]: uhh can I ignore this?

[the remark this answers is not in the document]
Comment (toyhammered) [reply lib/html/pipeline/gif_filter.rb]: nope. you can use `\1` though or just make it a group

[the remark this answers is not in the document]
Comment (toyhammered) [reply lib/html/pipeline/gif_filter.rb]: "https://gifs.com/gif/\1" should work

[the remark this answers is not in the document]
Comment (matthewdias) [reply lib/html/pipeline/gif_filter.rb]: 👍  ok i'll do that tomorrow

**Comment by matthewdias** [comment]
probably shouldn't merge this until https://github.com/discourse/onebox/pull/336

**Comment by toyhammered** [comment]
Also you might want to write some tests for this (you can use a fixture if you want to just copy/paste some html doc). 

cc @NuckChorris (incase he disagree's with me/has another solution)

**Comment by NuckChorris** [comment]
Is there a specific reason this isn't implemented as a `Onebox::Engine`? 🤔 

I mean, this code seems solid, I'm just a little confused by the choice of how you implemented it

**Comment by matthewdias** [comment]
@NuckChorris I figured that for onebox, they probably didn't want to make the choice to automatically convert direct links to files. None of the existing engines do that. As for just having custom engines on the server, I guess that would work, but I'd just be rewriting code that's already in onebox. If you think it would be better that way, I can do that instead.

**Comment by NuckChorris** [comment]
In theory couldn't you subclass the ones that are already in Onebox, so that you don't have to rewrite code?

**Comment by matthewdias** [comment]
ahhh yeah i could, that would be way better 😄

**Comment by NuckChorris** [comment]
If you do, I think `lib/onebox/engines` would be the right place for that, I wanna say we have some in V2 which did that

**Comment by matthewdias** [comment]
aye aye

**Comment by JoshFabian** [comment]
@matthewdias Is this something we'll be able to reuse once we begin hosting our own video / images? (very near future)

**Comment by NuckChorris** [comment]
@JoshFabian Not really, this just coerces things like imgur gif embeds into imgur video embeds by changing the URL

When we deal with hosting video/images, it's up to us to transcode them from GIF->WebM/MP4

**Comment by matthewdias** [comment]
I would advise just storing the videos when we do that.

**Comment by NuckChorris** [comment]
I'd probably say when we do that: store the videos, convert gifs to videos, make sure we have at least MP4, maybe WebM (but maybe not, as much as I personally support WebM I don't know of any browsers without MP4 support) 🤔

**Comment by matthewdias** [comment]
yeah, the advantage of webm is higher quality at a lower filesize, but if we have to store mp4 anyway, theres not much point. unless storage isn't an issue?

**Comment by wopian** [comment]
It'd be great when Safari decides to be a modern browser again and stops being the only browser to not support WebM 😢

**Comment by matthewdias** [comment]
@NuckChorris is there any way i can disable these [requires](https://github.com/discourse/onebox/blob/master/lib/onebox/engine.rb) so i can subclass the engines without getting circular require warnings? Works fine right now, there is just a lot of duplicate code.

**Comment by NuckChorris** [comment]
Perhaps requiring `onebox` manually before you do the subclasses will fix it?

**Comment by matthewdias** [comment]
i'll give that a try 👍

**Pull request closed without merging**